### PR TITLE
 Add CLI option to listen to stdin to close when the channel closes

### PR DESCRIPTION
**Pull request by LostKobrakai**
This PR contains:
- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?
- [ ] yes (*bugfixes and features will not be merged without tests*)
- [x] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:

### Description

Fixes https://github.com/rollup/rollup-watch/pull/57 again.

### CODE REVIEW

File: bin/src/run/watch.ts [hunk 1]
```diff
@@ -156,8 +156,9 @@ export default function watch(
 	process.on('uncaughtException', close);
 
 	// only listen to stdin if it is a pipe
-	if (!process.stdin.isTTY) {
+	if (!process.stdin.isTTY || command.watchStdin) {
 		process.stdin.on('end', close); // in case we ever support stdin!
+		process.stdin.resume();
```
Comment (evocateur) [inline bin/src/run/watch.ts]: This line looks like the only relevant fix?

It's not clear to me why we need `--watch-stdin`. If `rollup` is called directly, `process.stdin` is already a flowing stream, and moreover `Ctrl+C` is handled by `onExit`, not this conditional. What use case does `--watch-stdin` allow?

**Comment by lukastaegert** [comment]
I must admit I am not so deep into these kinds of things but this behaviour was changed because of issues with Lerna, specifically here: #2410, which was meant as a fix for #1919 (which was never closed because of missing feedback but I assume it worked as otherwise there was no further feedback).

> most other tools out there have a custom cli flag for that kind of functionality
Could you give some examples here and how they designed the API? My hope is that users should not need to worry too much about these things and Rollup just "does the right thing", but apparently this is difficult.

Seeing that we are playing some code ping pong here, I would really hope to have some kind of test that actually triggers the core issue here. Shouldn't this be possible in some way? Otherwise in half a year, someone will just again break this again.

I would also be interested to get some clarification for @evocateur 's comment.

File: bin/src/run/watch.ts (hunk 1 above)
Comment (evocateur) [reply bin/src/run/watch.ts]: If that's the case, it should be the only change in the PR.

**Comment by evocateur** [comment]
At a higher level, I'm not even sure why `rollup -w` even _cares_ about `stdin`. There's no API that I'm aware of that needs, supports, or consumes _anything_ from `stdin`—that is, keyboard or file input from an interactive terminal. It just doesn't make any sense to care until there's actually an API need for it.

Just remove the whole block. If nothing is being done with `stdin`, don't touch it. [on-exit](https://www.npmjs.com/package/on-exit) will handle everything else.
```patch
 	const removeOnExit = onExit(close);
	process.on('uncaughtException', close);

-	// only listen to stdin if it is a pipe
-	if (!process.stdin.isTTY) {
-		process.stdin.on('end', close); // in case we ever support stdin!
-	}

	function close(err: Error) {
		removeOnExit();
		process.removeListener('uncaughtException', close);
-		// removing a non-existent listener is a no-op
-		process.stdin.removeListener('end', close);

		screen.close();
		if (watcher) watcher.close();
```

**Comment by LostKobrakai** [comment]
That wouldn't solve my issue though. I'm using a framework, which starts rollup using erlang ports. Erlang ports communicate through stdin/stdout and expect the external program to exit if stdin is closed. Without listening on stdin rollup will never exit when I stop the erlang vm leaving an orphan process alive. I'm also not sure how on-exit would help here, as my issue is that the node process won't exit, not that I need to run something when it indeed does.

**Comment by evocateur** [comment]
That framework should be sending [standard signals](https://unix.stackexchange.com/questions/317492/list-of-kill-signals) to the subprocesses, not some weird `stdio` pattern. Send `SIGTERM` to the watch process and `on-exit` will handle it correctly.

**Comment by LostKobrakai** [comment]
If it would be possible natively I'd do that, but the erlang vm (as well as the jvm from what I read) don't expose the possibility to send signals to external processes. Also there's still the possibility of the parent process exiting without sending the necessary signals to clean up, while stdin will be closed by the OS if the parent exits. So I wouldn't consider common behaviour of many unix tools that weird.

**Comment by LostKobrakai** [comment]
I mean there's a reason why it's behind a feature flag in brunch and webpack. Because some people expect this to work like that and others don't. But really it's 2 loc and some boilerplate for having the cli flag.

**Comment by lukastaegert** [comment]
Ok, so I understand now that it makes sense to choose the feature flag as you did. Still what is keeping me from merging is the absence of a test and the fact that we already broke this once.

**Comment by Arno500** [comment]
This is a really needed feature, especially when having packages that depends on other to build. We can't run our app if rollup doesn't "exit" or resume the terminal after requirements are built.

**Comment by halostatue** [comment]
This would be extremely useful to have in place, as the [suggested alternative](https://hexdocs.pm/elixir/Port.html#module-zombie-os-processes) from Elixir is a wrapper script that suppresses the stderr output of the watcher, reducing its utility:

```bash
#! /usr/bin/env bash

declare -i watcher

"${@}" &
watcher=$!

while read -r; do
  :
done

kill $watcher
```

**Comment by shellscape** [comment]
@LostKobrakai it looks like this PR needs to be updated to master and a few conflicts resolved. Is this something you're up for, or should we nuke it and start a new PR?

**Comment by LostKobrakai** [comment]
There's so few changes I doubt updating is the quicker path, compared to just adding it to the current code.

**Comment by shellscape** [comment]
OK cool. Would you like to make those additions in a new PR so your contribution is retained? If not, it's no problem, we can tag you in the commit if you'd prefer.

**Comment by LostKobrakai** [comment]
I'd be fine without it. I just care for the functionality :)

**Comment by shellscape** [comment]
Closing for now as stale. @LostKobrakai if you'd like to reopen these please ping and it shall be done.

**Pull request closed without merging**